### ISSUE TITLE
[23] Type compatibility error in primitive type pattern

### ISSUE THREAD

**Issue by jarthana** [eclipse-jdt/eclipse.jdt.core#2891]
Given the test code:
```
record RforRecord(long x) {}
public class X  {
    public void foo() {
        if (new RforRecord(0L) instanceof RforRecord(int y)) {
            System.out.println("Yay");
        }
    }
}
```
ECJ reports an error `"Record component with type long is not compatible with type int"` and this is accepted by Javac.

**Comment by jarthana** [comment]
Copying @mpalat  and @stephan-herrmann who might be interested in this.

**Comment by stephan-herrmann** [comment]
> ECJ reports an error `"Record component with type long is not compatible with type int"` and this is accepted by Javac.

With latest in BETA_JAVA23 resolution passes, but then we get "Unexpected operand at stack top" from `OperandStack.pop()`.

Investigating.

**Comment by stephan-herrmann** [comment]
Thanks @jarthana, your test demonstrated that recent changes for code gen where not at all wired into type patterns nested in a record pattern.

Should now work with #2894 but we could use way more tests, so be my guests :)

**Comment by stephan-herrmann** [comment]
Fixed via #2894
